Makefile

```diff
@@ -78,48 +113,95 @@ version:
 	     -ldflags '-X $(PKG)/pkg/version.version=$(VERSION)' \
 	     $(BUILD_TAGS) cmd/logcounter/log_counter.go
 
 ./bin/node-problem-detector: $(PKG_SOURCES)
-	CGO_ENABLED=$(CGO_ENABLED) GOOS=linux go build -o bin/node-problem-detector \
+	CGO_ENABLED=$(CGO_ENABLED) GOOS=linux GOARCH=$(ARCH) go build -o bin/node-problem-detector-$(ARCH) \
 	     -ldflags '-X $(PKG)/pkg/version.version=$(VERSION)' \
 	     $(BUILD_TAGS) cmd/node_problem_detector.go
 
+qemu-register:
+	docker run --rm --privileged multiarch/qemu-user-static:register --reset
+
 Dockerfile: Dockerfile.in
 	sed -e 's|@BASEIMAGE@|$(BASEIMAGE)|g' $< >$@
+	sed -i "s|-ARCH|-$(QEMUARCH)|g" $@
+	sed -i "s|-BINARCH|-$(ARCH)|g" $@
+ifeq ($(ARCH),amd64)
+	# When building "normally" for amd64, remove the whole line, it has no part in the amd64 image
+	sed -i "/CROSS_BUILD_/d" Dockerfile
+else
+	curl -sSL https://github.com/multiarch/qemu-user-static/releases/download/$(QEMUVERSION)/x86_64_qemu-$(QEMUARCH)-static.tar.gz | tar -xz -C .
+	sed -i "s/CROSS_BUILD_//g" Dockerfile
+endif
 
 test: vet fmt
 	go test -timeout=1m -v -race ./cmd/options ./pkg/... $(BUILD_TAGS)
 
 build-binaries: ./bin/node-problem-detector ./bin/log-counter
 
-build-container: build-binaries Dockerfile
-	docker build -t $(IMAGE) .
+build-all-container: qemu-register $(addprefix build-container-,$(ALL_ARCH))
 
-build-tar: ./bin/node-problem-detector ./bin/log-counter
+build-container-%:
+	$(MAKE) ARCH=$* build-container
+
+build-container: clean build-in-docker Dockerfile
+	docker build -t $(IMAGE_WITH_ARCH) .
+
+build-tar: build-in-docker
+	cp bin/node-problem-detector-amd64 bin/node-problem-detector
+	cp bin/log-counter-amd64 bin/log-counter
 	tar -zcvf $(TARBALL) bin/ config/
 	sha1sum $(TARBALL)
 	md5sum $(TARBALL)
 
+build-tar-all:
+	@for arch in $(ALL_ARCH); do $(MAKE) ARCH=$${arch} build-in-docker; done
+	cp bin/node-problem-detector-amd64 bin/node-problem-detector
+	cp bin/log-counter-amd64 bin/log-counter
+	tar -zcvf $(TARBALL-ALL) bin/ config/
+	sha1sum $(TARBALL-ALL)
+	md5sum $(TARBALL-ALL)
+
+build-all: build-all-container build-tar-all
+
 build: build-container build-tar
 
 docker-builder:
 	docker build -t npd-builder ./builder
 
-build-in-docker: clean docker-builder
-	docker run -v `pwd`:/gopath/src/k8s.io/node-problem-detector/ npd-builder:latest bash -c 'cd /gopath/src/k8s.io/node-problem-detector/ && make build-binaries'
+build-in-docker: docker-builder
+	docker run -e CC=$(CC) -e GOARM=$(GOARM) -e GOARCH=$(ARCH) \
+		-u $(shell id -u ${USER}):$(shell id -g ${USER}) \
+		-v `pwd`:/gopath/src/k8s.io/node-problem-detector/ npd-builder:latest bash -c \
+		'cd /gopath/src/k8s.io/node-problem-detector/ && make build-binaries'
+
+push-all-container: $(addprefix push-container-,$(ALL_ARCH))
+
+push-container-%:
+	$(MAKE) ARCH=$* push-container
 
 push-container: build-container
-	gcloud docker -- push $(IMAGE)
+	docker push $(IMAGE_WITH_ARCH)
```

Comment:
I think you need to use `gcloud docker -- push ...` here. You got errors pushing containers in the tests.

And why is `IMAGE_WITH_ARCH` used instead of `IMAGE`?

Reply:
> I think you need to use `gcloud docker -- push ...` here. You got errors pushing containers in the tests.

I can modify this step but we need to find some other way to make `docker` command work because we need to run `docker manifest` command to run successfully to push the manifest to gcr.

I know that most of the folks use docker command to push images to gcr which works fine, @ixdy do you have any idea what configuration you change to make it work.

> 
> And why is `IMAGE_WITH_ARCH` used instead of `IMAGE`?

this is because we want to push all the images with architecture first and latest `IMAGE` will be the fat manifest image which will be referring to all the `IMAGE_WITH_ARCH` images.

Reply:
`gcloud docker` has been deprecated for a while.

Reply:
OK. Then you need `gcloud auth configure-docker` before pushing. I just merged https://github.com/kubernetes/node-problem-detector/pull/263 to fix the deprecation issue. You can rebase.